File: android/src/main/java/com/backgroundlocation/service/SyncService.java
package com.backgroundlocation.service;

import android.content.Context;
import android.os.AsyncTask;

import com.backgroundlocation.config.Config;
import com.backgroundlocation.data.LocationModel;
import com.backgroundlocation.data.sqlite.SQLiteLocationDAO;
import com.backgroundlocation.event.HttpResponseEvent;
import com.backgroundlocation.util.LogHelper;

import org.greenrobot.eventbus.EventBus;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * HTTP Sync Service
 * LOCKING mekanizmalı batch sync
 * 
 * CRITICAL FEATURES:
 * 1. Locking: kayıtları sync sırasında kilitler
 * 2. Batch: toplu gönderim (multi-location)
 * 3. Unlock: başarısız olursa kilidi açar
 * 4. Destroy: başarılı olursa siler
 * 5. Retry: offline iken queue'da bekler
 */
public class SyncService {
    
    private static final String TAG = "SyncService";
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
    
    // Thread-safe busy flag ()
    private static final AtomicBoolean isSyncing = new AtomicBoolean(false);
    
    /**
     * Sync locations to server with LOCKING mechanism
     * AtomicBoolean compareAndSet
     */
    public static void sync(Context context) {
        // Thread-safe check ()
        if (isSyncing.compareAndSet(false, true)) {
            new SyncTask(context).execute();
        } else {
            LogHelper.d(context, TAG, "⏸️ HttpService is busy, skipping");
        }
    }
    
    private static class SyncTask extends AsyncTask<Void, Void, SyncResult> {
        private Context context;
        private Config config;
        private SQLiteLocationDAO database;
        
        public SyncTask(Context context) {
            this.context = context.getApplicationContext();
            this.config = Config.getInstance(this.context);
            this.database = SQLiteLocationDAO.getInstance(this.context);
        }
        
        @Override
        protected void onPreExecute() {
            // isSyncing already set in sync() method
        }
        
        @Override
        protected SyncResult doInBackground(Void... voids) {
            try {
                // 1. Get unlocked count ()
                int unlockedCount = database.count(true); // only unlocked
                
                LogHelper.d(TAG, "📊 Unlocked locations: " + unlockedCount);
                
                // Check threshold
                if (config.autoSyncThreshold > 0 && unlockedCount < config.autoSyncThreshold) {
                    android.util.Log.i(TAG, "⏸️ Below threshold (" + config.autoSyncThreshold + "), skipping sync");
                    return new SyncResult(true, 0, "Below threshold");
                }
                
                // 2. CRITICAL: Get locations WITH LOCKING ()
                // This will SELECT WHERE locked=0 and UPDATE SET locked=1
                List<LocationModel> locations = database.allWithLocking(config.maxBatchSize);
                
                if (locations.isEmpty()) {
                    LogHelper.d(TAG, "ℹ️ No locations to sync");
                    return new SyncResult(true, 0, "No locations to sync");
                }
                
                LogHelper.d(TAG, "🔒 Locked " + locations.size() + " records (allWithLocking)");
                
                // 4. Convert to JSON array (batch)
                JSONArray jsonArray = new JSONArray();
                for (LocationModel location : locations) {
                    jsonArray.put(location.toJSON());
                }
                
                // 5. Prepare request body
                JSONObject body = new JSONObject();
                
                // Check batchSync mode
                if (config.batchSync) {
                    // Batch mode: wrap in "locations" array
                    body.put("locations", jsonArray);
                } else {
                    // Single mode: send first location only
                    if (jsonArray.length() > 0) {
                        body = jsonArray.getJSONObject(0);
                    }
                }
                
                // Merge params from config
                if (config.params != null && !config.params.isEmpty()) {
                    try {
                        JSONObject paramsJson = new JSONObject(config.params);
                        Iterator<String> keys = paramsJson.keys();
                        while (keys.hasNext()) {
                            String key = keys.next();
                            body.put(key, paramsJson.get(key));
                        }
                    } catch (Exception e) {
                        LogHelper.w(TAG, "Failed to merge params: " + e.getMessage());
                    }
                }
                
                // 6. Create HTTP client
                OkHttpClient client = new OkHttpClient.Builder()
                        .connectTimeout(30, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
                        .writeTimeout(30, TimeUnit.SECONDS)
                        .build();
                
                // 7. Build request with headers
                RequestBody requestBody = RequestBody.create(body.toString(), JSON);
                Request.Builder requestBuilder = new Request.Builder()
                        .url(config.url);
                
                // Add headers from config
                requestBuilder.addHeader("Content-Type", "application/json");
                if (config.headers != null && !config.headers.isEmpty()) {
                    try {
                        JSONObject headersJson = new JSONObject(config.headers);
                        Iterator<String> keys = headersJson.keys();
                        while (keys.hasNext()) {
                            String key = keys.next();
                            String value = headersJson.getString(key);
                            requestBuilder.addHeader(key, value);
                        }
                    } catch (Exception e) {
                        LogHelper.w(TAG, "Failed to add headers: " + e.getMessage());
                    }
                }
                
                if (config.method.equalsIgnoreCase("POST")) {
                    requestBuilder.post(requestBody);
                } else if (config.method.equalsIgnoreCase("PUT")) {
                    requestBuilder.put(requestBody);
                }
                
                Request request = requestBuilder.build();
                
                LogHelper.d(TAG, "HTTP " + config.method + " batch (" + locations.size() + ") to " + config.url);
                
                // 8. Execute request
                Response response = client.newCall(request).execute();
                int statusCode = response.code();
                String responseBody = response.body() != null ? response.body().string() : "";
                
                boolean success = response.isSuccessful();
                
                LogHelper.d(TAG, "HTTP Response: " + statusCode + " - " +
                    (success ? "SUCCESS" : "FAILED"));
                
                // 9. Emit HTTP event (direct EventBus)
                HttpResponseEvent httpEvent = new HttpResponseEvent(statusCode, success, responseBody);
                EventBus.getDefault().post(httpEvent);
                
                if (success) {
                    // 10. SUCCESS: Delete synced locations ()
                    database.destroyAll(locations);
                    LogHelper.d(TAG, "✅ DELETED " + locations.size() + " synced records");
                    
                    // Check if there are more to sync (recursive pattern)
                    int remaining = database.count(true); // unlocked count
                    if (config.autoSyncThreshold > 0 && remaining >= config.autoSyncThreshold) {
                        // Recursively sync more
                        LogHelper.d(TAG, "🔄 More locations to sync (" + remaining + "), continuing...");
                        return doInBackground(); // Recursive call
                    }
                    
                    return new SyncResult(true, locations.size(), responseBody);
                } else {
                    // 11. FAILURE: Unlock locations for retry ()
                    database.unlock(locations);
                    LogHelper.w(TAG, "🔓 UNLOCKED " + locations.size() + " records (will retry later)");
                    
                    return new SyncResult(false, 0, "HTTP " + statusCode + ": " + responseBody);
                }
                
            } catch (Exception e) {
                e.printStackTrace();
                
                // On exception, unlock all to be safe ()
                try {
                    database.unlockAll();
                    LogHelper.w(TAG, "🔓 Unlocked all locations due to exception");
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
                
                return new SyncResult(false, 0, e.getMessage());
            }
        }
        
        @Override
        protected void onPostExecute(SyncResult result) {
            // Release busy flag ()
            isSyncing.set(false);
            
            if (!result.success && result.message != null) {
                LogHelper.e(TAG, "❌ Sync failed: " + result.message);
            } else if (result.success && result.count > 0) {
                LogHelper.d(TAG, "✅ Sync completed: " + result.count + " locations synced");
            }
        }
    }
    
    /**
     * Result class for sync operation
     */
    private static class SyncResult {
        boolean success;
        int count;
        String message;
        
        SyncResult(boolean success, int count, String message) {
            this.success = success;
            this.count = count;
            this.message = message;
        }
    }
    
    /**
     * Check if network is available
     */
    public static boolean isNetworkAvailable(Context context) {
        try {
            android.net.ConnectivityManager cm = 
                (android.net.ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            android.net.NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
            return activeNetwork != null && activeNetwork.isConnectedOrConnecting();
        } catch (Exception e) {
            return false;
        }
    }
}

File: ios/service/SyncService.m
//
//  SyncService.m
//  RNBackgroundLocation
//
//  HTTP Sync Service
//  Android SyncService.java benzeri
//   LOCKING mekanizmalı batch sync
//

#import "SyncService.h"
#import "TSConfig.h"
#import "TSAuthorization.h"
#import "LocationModel.h"
#import "SQLiteLocationDAO.h"
#import "HttpResponseEvent.h"
#import "LogHelper.h"
#import "TSReachability.h"

@interface SyncService ()
@property (nonatomic, strong) TSConfig *config;
@property (nonatomic, strong) SQLiteLocationDAO *database;
@property (nonatomic, assign) BOOL isSyncing; // AtomicBoolean benzeri (iOS'ta @synchronized kullanacağız)
@property (nonatomic, assign) BOOL isRecursiveSync; // Flag to skip threshold check in recursive calls
@end

@implementation SyncService

+ (instancetype)sharedInstance {
    static SyncService *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[SyncService alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _config = [TSConfig sharedInstance];
        _database = [SQLiteLocationDAO sharedInstance];
        _isSyncing = NO;
    }
    return self;
}

+ (void)sync {
    [[self sharedInstance] sync];
}

- (void)sync {
    // Thread-safe check (Android AtomicBoolean.compareAndSet benzeri)
    @synchronized(self) {
        if (self.isSyncing) {
            [LogHelper i:@"SyncService" message:@"⏸️ HttpService is busy, skipping"];
            return;
        }
        self.isSyncing = YES;
    }
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self performSync];
    });
}

- (void)performSync {
    @try {
        // CRITICAL: Check network connection FIRST
        // If no internet, don't try to sync - locations are already saved in SQLite
        TSReachability *reachability = [TSReachability reachabilityForInternetConnection];
        BOOL hasNetwork = [reachability isReachable];
        
        if (!hasNetwork) {
            [LogHelper w:@"SyncService" message:@"📵 No network connection, skipping sync (locations saved in SQLite for later)"];
            [self releaseSync];
            return;
        }
        
        // 1. Get unlocked count ()
        NSInteger unlockedCount = [self.database countOnlyUnlocked:YES];
        
        [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"📊 Unlocked locations: %ld (recursive: %@)", (long)unlockedCount, self.isRecursiveSync ? @"YES" : @"NO"]];
        
        // CRITICAL: Check threshold ONLY for initial sync trigger (NOT in recursive calls)
        // If autoSyncThreshold > 0, only start sync if count >= threshold
        // BUT: Once sync starts (recursive), continue until ALL locations are synced
        // This ensures offline locations are all synced when internet returns
        // CRITICAL: Skip threshold check if this is a recursive sync call
        if (!self.isRecursiveSync && self.config.autoSyncThreshold > 0 && unlockedCount < self.config.autoSyncThreshold) {
            [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"⏸️ Below threshold (%ld < %ld), skipping sync", (long)unlockedCount, (long)self.config.autoSyncThreshold]];
            [self releaseSync];
            return;
        }
        
        // If we have locations to sync (even if below threshold in recursive calls), continue
        if (unlockedCount == 0) {
            [LogHelper i:@"SyncService" message:@"ℹ️ No locations to sync"];
            [self releaseSync];
            return;
        }
        
        // 2. CRITICAL: Get locations WITH LOCKING ()
        // This will SELECT WHERE locked=0 and UPDATE SET locked=1
        NSArray<LocationModel *> *locations = [self.database allWithLocking:self.config.maxBatchSize];
        
        if (locations.count == 0) {
            [LogHelper i:@"SyncService" message:@"ℹ️ No locations to sync"];
            [self releaseSync];
            return;
        }
        
        [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"🔒 Locked %lu records (allWithLocking)", (unsigned long)locations.count]];
        
        // 3. Convert to JSON array (batch)
        NSMutableArray *jsonArray = [NSMutableArray array];
        for (LocationModel *location in locations) {
            [jsonArray addObject:[location toDictionary]];
        }
        
        // 4. Prepare request body
        NSMutableDictionary *body = [NSMutableDictionary dictionary];
        
        // Check batchSync mode
        if (self.config.batchSync) {
            // Batch mode: wrap in "locations" array
            body[@"locations"] = jsonArray;
        } else {
            // Single mode: send first location only
            if (jsonArray.count > 0) {
                body = [jsonArray.firstObject mutableCopy];
            }
        }
        
        // Merge params from config
        NSDictionary *params = [self.config getParamsDictionary];
        if (params.count > 0) {
            [body addEntriesFromDictionary:params];
        }
        
        // 5. Create URL request
        NSURL *url = [NSURL URLWithString:self.config.url];
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
        request.HTTPMethod = self.config.method ?: @"POST";
        [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
        
        // Add headers from config
        NSDictionary *headers = [self.config getHeadersDictionary];
        for (NSString *key in headers) {
            [request setValue:headers[key] forHTTPHeaderField:key];
        }
        
        // Apply authorization (TSAuthorization pattern)
        if (self.config.authorization) {
            [self.config.authorization apply:request];
        }
        
        // Body
        NSError *error;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:body options:0 error:&error];
        if (error) {
            [LogHelper e:@"SyncService" message:[NSString stringWithFormat:@"❌ Failed to serialize JSON: %@", error.localizedDescription] error:error];
            [self unlockLocations:locations];
            [self releaseSync];
            return;
        }
        request.HTTPBody = jsonData;
        
        [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"HTTP %@ batch (%lu) to %@", self.config.method, (unsigned long)locations.count, self.config.url]];
        
        // 6. Execute request
        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
        config.timeoutIntervalForRequest = 30.0;
        config.timeoutIntervalForResource = 30.0;
        
        NSURLSession *session = [NSURLSession sessionWithConfiguration:config];
        NSURLSessionDataTask *task = [session dataTaskWithRequest:request 
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
            NSInteger statusCode = httpResponse.statusCode;
            NSString *responseText = data ? [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] : @"";
            
            BOOL success = (statusCode >= 200 && statusCode < 300) && (error == nil);
            
            [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"HTTP Response: %ld - %@", (long)statusCode, success ? @"SUCCESS" : @"FAILED"]];
            
            // 7. Emit HTTP event (Android EventBus yerine callback)
            HttpResponseEvent *httpEvent = [[HttpResponseEvent alloc] initWithStatusCode:statusCode 
                                                                                  success:success 
                                                                            responseText:responseText];
            if (self.onHttpCallback) {
                self.onHttpCallback(httpEvent);
            }
            
            if (success) {
                // 8. SUCCESS: Delete synced locations ()
                [self.database destroyAll:locations];
                [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"✅ DELETED %lu synced records", (unsigned long)locations.count]];
                
                // CRITICAL: Check if there are more to sync (Transistorsoft recursive pattern)
                // Recursive sync should continue until ALL pending locations are synced
                // Don't check threshold here - if there are any remaining locations, sync them
                NSInteger remaining = [self.database countOnlyUnlocked:YES];
                if (remaining > 0) {
                    // Recursively sync more - continue until all locations are synced
                    [LogHelper i:@"SyncService" message:[NSString stringWithFormat:@"🔄 More locations to sync (%ld), continuing recursive sync...", (long)remaining]];
                    // CRITICAL: Set recursive flag BEFORE recursive call
                    self.isRecursiveSync = YES;
                    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                        [self performSync]; // Recursive call - will sync ALL remaining locations (threshold check skipped)
                    });
                    return; // Don't release sync yet, continue syncing
                } else {
                    [LogHelper i:@"SyncService" message:@"✅ All locations synced successfully"];
                    // CRITICAL: Reset recursive flag when all locations are synced
                    self.isRecursiveSync = NO;
                }
            } else {
                // 9. FAILURE: Unlock locations for retry ()
                [self unlockLocations:locations];
                [LogHelper w:@"SyncService" message:[NSString stringWithFormat:@"🔓 UNLOCKED %lu records (will retry later)", (unsigned long)locations.count]];
            }
            
            [self releaseSync];
        }];
        
        [task resume];
        
    } @catch (NSException *exception) {
        [LogHelper e:@"SyncService" message:[NSString stringWithFormat:@"❌ Sync exception: %@", exception.reason]];
        
        // On exception, unlock all to be safe ()
        @try {
            [self.database unlockAll];
            [LogHelper w:@"SyncService" message:@"🔓 Unlocked all locations due to exception"];
        } @catch (NSException *ex) {
            [LogHelper e:@"SyncService" message:[NSString stringWithFormat:@"❌ Failed to unlock: %@", ex.reason]];
        }
        
        [self releaseSync];
    }
}

- (void)unlockLocations:(NSArray<LocationModel *> *)locations {
    [self.database unlock:locations];
}

- (void)releaseSync {
    @synchronized(self) {
        self.isSyncing = NO;
        // CRITICAL: Reset recursive flag when sync is released
        // This ensures next sync starts fresh (with threshold check)
        self.isRecursiveSync = NO;
    }
}

@end

